ip_country/src/country_block_stream.rs

```diff
@@ -0,0 +1,411 @@
+                    Ordering::Less
+                } else {
+                    Ordering::Equal
+                }
+            }
+            (IpAddr::V6(ip), IpRange::V6(low, hi)) => {
+                let (low_range, hi_range) = (u128::from(*low), u128::from(*hi));
+                let ip_num = u128::from(ip);
+                if ip_num < low_range {
+                    Ordering::Greater
+                } else if ip_num > hi_range {
+                    Ordering::Less
+                } else {
+                    Ordering::Equal
+                }
+            }
+            (_ip, _range) => panic!("Mismatch ip and range versions"), //TODO write test for this panic!
+        }
+    }
+
+    fn contains_inner(begin: u128, end: u128, candidate: u128) -> bool {
+        (candidate >= begin) && (candidate <= end)
+    }
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct CountryBlock {
+    pub ip_range: IpRange,
+    pub country: Country,
+}
+
+impl TryFrom<StringRecord> for CountryBlock {
+    type Error = String;
+
+    fn try_from(string_record: StringRecord) -> Result<Self, Self::Error> {
+        let mut iter = string_record.iter();
+        let start_ip = Self::ip_addr_from_iter(&mut iter)?;
+        let end_ip = Self::ip_addr_from_iter(&mut iter)?;
+        let iso3166 = match iter.next() {
+            None => return Err("CSV line contains no ISO 3166 country code".to_string()),
+            Some(s) => s,
+        };
+        if iter.next().is_some() {
+            return Err(format!(
+                "CSV line should contain 3 elements, but contains {}",
+                string_record.len()
+            ));
+        };
+        Self::validate_ip_addresses(start_ip, end_ip)?;
+        let country = Country::try_from(iso3166)?;
+        let country_block = match (start_ip, end_ip) {
+            (IpAddr::V4(start), IpAddr::V4(end)) => CountryBlock {
+                ip_range: IpRange::V4(start, end),
+                country,
+            },
+            (IpAddr::V6(start), IpAddr::V6(end)) => CountryBlock {
+                ip_range: IpRange::V6(start, end),
+                country,
+            },
+            (start, end) => panic!(
+                "Start and end addresses must be of the same type, not {} and {}",
+                start, end
+            ),
+        };
+        Ok(country_block)
+    }
+}
+
+impl CountryBlock {
+    fn ip_addr_from_iter(iter: &mut StringRecordIter) -> Result<IpAddr, String> {
+        let ip_string = match iter.next() {
+            None => return Err("Missing IP address in CSV record".to_string()),
+            Some(s) => s,
+        };
+        let ip_addr = match IpAddr::from_str(ip_string) {
+            Err(e) => {
+                return Err(format!(
+                    "Invalid ({:?}) IP address in CSV record: '{}'",
+                    e, ip_string
+                ))
+            }
+            Ok(ip) => ip,
+        };
+        Ok(ip_addr)
```

Comment:
You can rewrite this without the second return clause. Just don't create the `ip_addr` variable and return implicitly by leaving the match stm.